File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/GameObject.java
package com.iamagamedev.mybrandnewgame.gameObjects;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Rect;

import com.iamagamedev.mybrandnewgame.Animation;
import com.iamagamedev.mybrandnewgame.LocationXYZ;
import com.iamagamedev.mybrandnewgame.RectHitBox;

import java.util.ArrayList;

/**
 * Created by Михан on 24.05.2017.
 */

public abstract class GameObject {

    public final int LEFT = -1;
    public final int RIGHT = 1;
    public final int UP = -2;
    public final int DOWN = 2;

    private LocationXYZ worldLocation;
    private RectHitBox rectHitBoxBottom = new RectHitBox();
    private RectHitBox rectHitBoxLeft = new RectHitBox();
    private RectHitBox rectHitBoxTop = new RectHitBox();
    private RectHitBox rectHitBoxRight = new RectHitBox();
    private Animation animation = null;
    private boolean animated;
    private int animFps = 1;
    private boolean moves = false;
    private int facing;

    private float width;
    private float height;

    private float health;
    private float damage;

    private float xVelocity;
    private float yVelocity;

    private boolean active = true;
    private boolean visible = true;
    private int animFrameCount = 1;
    private char type;

    private String bitmapName;
    private String badBitmapName;
    public float speed;
    private boolean isTalking;
    private ArrayList<String> dialogs;
    private boolean canTalk;


    public abstract void update(long fps);

    public Bitmap prepareBitmap(Context context,
                                String bitmapName, int pixelsPerMetre) {
        int resID = context.getResources()
                .getIdentifier(bitmapName,
                        "drawable", context.getPackageName());

        Bitmap bitmap = BitmapFactory.decodeResource(context.getResources(), resID);
        bitmap = Bitmap.createScaledBitmap(bitmap, (int) (width * animFrameCount * pixelsPerMetre),
                (int) (height * pixelsPerMetre), false);

        return bitmap;
    }

    public LocationXYZ getWorldLocation() {
        return worldLocation;
    }

    public void setWorldLocation(float x, float y, int z) {
        this.worldLocation = new LocationXYZ();
        this.worldLocation.x = x;
        this.worldLocation.y = y;
        this.worldLocation.z = z;
    }

    public void setRectHitBox() {
        LocationXYZ locationXYZ = getWorldLocation();
        float goX = locationXYZ.x;
        float goY = locationXYZ.y;
        //update the obj feet hitbox
        rectHitBoxBottom.top = goY + getHeight() * .95f;
        rectHitBoxBottom.left = goX + getWidth() * .2f;
        rectHitBoxBottom.bottom = goY + getHeight() * .98f;
        rectHitBoxBottom.right = goX + getWidth() * .8f;
        // Update obj head hitbox
        rectHitBoxTop.top = goY;
        rectHitBoxTop.left = goX + getWidth() * .3f;
        rectHitBoxTop.bottom = goY + getHeight() * .2f;
        rectHitBoxTop.right = goX + getWidth() * .7f;
        // Update obj left hitbox
        rectHitBoxLeft.top = goY + getHeight() * .2f;
        rectHitBoxLeft.left = goX + getWidth() * .2f;
        rectHitBoxLeft.bottom = goY + getHeight() * .8f;
        rectHitBoxLeft.right = goX + getWidth() * .3f;
        // Update obj right hitbox
        rectHitBoxRight.top = goY + getHeight() * .2f;
        rectHitBoxRight.left = goX + getWidth() * .71f;
        rectHitBoxRight.bottom = goY + getHeight() * .8f;
        rectHitBoxRight.right = goX + getWidth() * .8f;
    }

    public void setAnimated(int pixelsPerMetre, boolean animated) {
        this.animated = animated;
        this.animation = new Animation(bitmapName,
                height, width, animFps, animFrameCount, pixelsPerMetre);
    }

    public Rect getRectToDraw(long deltaTime) {
        return animation.getCurrentFrame(deltaTime, xVelocity, yVelocity, isMoves());
    }

    public void move(long fps) {
        if (xVelocity != 0) {
            this.getWorldLocation().x += xVelocity / fps;
        }
        if (yVelocity != 0) {
            this.getWorldLocation().y += yVelocity / fps;
        }
    }

    public int checkCollisions(RectHitBox rectHitBoxRight, RectHitBox rectHitBoxLeft,
                               RectHitBox rectHitBoxTop, RectHitBox rectHitBoxBottom) {
        int collided = 0;

        if (this.getRectHitBoxLeft().intersects(rectHitBoxRight)) {
            this.setWorldLocationX(rectHitBoxRight.right);
            collided = 1;
        }
        if (this.getRectHitBoxRight().intersects(rectHitBoxLeft)) {
            this.setWorldLocationX(rectHitBoxLeft.left - getWidth());
            collided = 1;
        }
        if (this.getRectHitBoxBottom().intersects(rectHitBoxTop)) {
            this.setWorldLocationY(rectHitBoxTop.top - getHeight());
            collided = 2;
        }
        if (this.getRectHitBoxTop().intersects(rectHitBoxBottom)) {
            this.setWorldLocationY(rectHitBoxBottom.bottom);
            collided = 2;
        }
        return collided;
    }

    public void setBitmapName(String bitmapName) {
        this.bitmapName = bitmapName;
    }

    public String getBitmapName() {
        return bitmapName;
    }

    public void setBadBitmapName(String badBitmapName) {
        this.badBitmapName = badBitmapName;
    }

    public String getBadBitmapName() {
        return badBitmapName;
    }

    public float getWidth() {
        return width;
    }

    public void setWidth(float width) {
        this.width = width;
    }

    public float getHeight() {
        return height;
    }

    public void setHeight(float height) {
        this.height = height;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public boolean isVisible() {
        return visible;
    }

    public void setVisible(boolean visible) {
        this.visible = visible;
    }

    public char getType() {
        return type;
    }

    public void setType(char type) {
        this.type = type;
    }

    public void setWorldLocationY(float y) {
        this.worldLocation.y = y;
    }

    public void setWorldLocationX(float x) {
        this.worldLocation.x = x;
    }

    public void setAnimFps(int animFps) {
        this.animFps = animFps;
    }

    public void setAnimFrameCount(int animFrameCount) {
        this.animFrameCount = animFrameCount;
    }

    public boolean isAnimated() {
        return animated;
    }

    public float getxVelocity() {
        return xVelocity;
    }

    public float getyVelocity() {
        return yVelocity;
    }

    public void setxVelocity(float xVelocity) {
        if (moves)
            this.xVelocity = xVelocity;
    }

    public void setyVelocity(float yVelocity) {
        if (moves)
            this.yVelocity = yVelocity;
    }

    public boolean isMoves() {
        return moves;
    }

    public void setMoves(boolean moves) {
        this.moves = moves;
    }

    public int getFacing() {
        return facing;
    }

    public void setFacing(int facing) {
        this.facing = facing;
    }

    public float getHealth() {
        return health;
    }

    public void setHealth(float health) {
        this.health = health;
    }

    public void setDamage(float damage) {
        this.damage = damage;
    }

    public float getDamage() {
        return damage;
    }

    public float getSpeed() {
        return speed;
    }

    public void setSpeed(float speed) {
        this.speed = speed;
    }

    public boolean isTalking() {
        return isTalking;
    }

    public void setTalking(boolean talking) {
        isTalking = talking;
    }

    public void setDialogs(ArrayList<String> dialogs) {
        this.dialogs = dialogs;
    }

    public ArrayList<String> getDialogs() {
        return dialogs;
    }

    public boolean isCanTalk() {
        return canTalk;
    }

    public void setCanTalk(boolean canTalk) {
        this.canTalk = canTalk;
    }

    public RectHitBox getRectHitBoxLeft() {
        return rectHitBoxLeft;
    }

    public RectHitBox getRectHitBoxBottom() {
        return rectHitBoxBottom;
    }

    public RectHitBox getRectHitBoxRight() {
        return rectHitBoxRight;
    }

    public RectHitBox getRectHitBoxTop() {
        return rectHitBoxTop;
    }
}

File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/GameObject.kt
package com.iamagamedev.mybrandnewgame.gameObjects

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Rect
import android.util.Log
import com.iamagamedev.mybrandnewgame.Animation
import com.iamagamedev.mybrandnewgame.LocationXYZ
import com.iamagamedev.mybrandnewgame.RectHitBox
import java.util.*

/**
 * Created by Михан on 24.05.2017.
 */
abstract class GameObject {
    @JvmField
    val LEFT = -1
    @JvmField
    val RIGHT = 1
    @JvmField
    val UP = -2
    @JvmField
    val DOWN = 2
    var worldLocation: LocationXYZ? = null
        private set
    val rectHitBoxBottom = RectHitBox()
    val rectHitBoxLeft = RectHitBox()
    val rectHitBoxTop = RectHitBox()
    val rectHitBoxRight = RectHitBox()
    private var animation: Animation? = null
    var isAnimated = false
        private set
    private var animFps = 1
    open var isMoves = false
    var facing = 0
    var width = 0f
    var height = 0f
    var health = 0
    var damage = 0
    private var xVelocity = 0f
    private var yVelocity = 0f
    var isActive = true
    var isVisible = true
    private var animFrameCount = 1
    var type = 0.toChar()
    var bitmapName: String? = null
    var badBitmapName: String? = null
    var speed = 0f
    var isTalking = false
    var dialogs: ArrayList<String>? = null
    var isCanTalk = false
    abstract fun update(fps: Long)
    fun prepareBitmap(context: Context,
                      bitmapName: String?, pixelsPerMetre: Int): Bitmap {
        val resID = context.resources
                .getIdentifier(bitmapName,
                        "drawable", context.packageName)
        var bitmap = BitmapFactory.decodeResource(context.resources, resID)
        bitmap = Bitmap.createScaledBitmap(bitmap, (width * animFrameCount * pixelsPerMetre).toInt(),
                (height * pixelsPerMetre).toInt(), false)
        return bitmap
    }

    fun setWorldLocation(x: Float, y: Float, z: Int) {
        worldLocation = LocationXYZ()
        worldLocation!!.x = x
        worldLocation!!.y = y
        worldLocation!!.z = z
    }

    open fun updatePosition(x: Float, y: Float) {
        setWorldLocation(x, y, worldLocation!!.z)
    }

    fun setRectHitBox() {
        val locationXYZ = worldLocation
        val goX = locationXYZ!!.x
        val goY = locationXYZ.y
        //update the obj feet hitbox
        rectHitBoxBottom.top = goY + height * .95f
        rectHitBoxBottom.left = goX + width * .1f
        rectHitBoxBottom.bottom = goY + height * .98f
        rectHitBoxBottom.right = goX + width * .9f
        // Update obj head hitbox
        rectHitBoxTop.top = goY
        rectHitBoxTop.left = goX + width * .1f
        rectHitBoxTop.bottom = goY + height * .2f
        rectHitBoxTop.right = goX + width * .9f
        // Update obj left hitbox
        rectHitBoxLeft.top = goY + height * .2f
        rectHitBoxLeft.left = goX + width * .2f
        rectHitBoxLeft.bottom = goY + height * .8f
        rectHitBoxLeft.right = goX + width * .3f
        // Update obj right hitbox
        rectHitBoxRight.top = goY + height * .2f
        rectHitBoxRight.left = goX + width * .71f
        rectHitBoxRight.bottom = goY + height * .8f
        rectHitBoxRight.right = goX + width * .8f
    }

    fun setAnimated(pixelsPerMetre: Int, animated: Boolean) {
        isAnimated = animated
        animation = Animation(bitmapName,
                height, width, animFps, animFrameCount, pixelsPerMetre)
    }

    fun getRectToDraw(deltaTime: Long): Rect {
        return animation!!.getCurrentFrame(deltaTime, xVelocity, yVelocity, isMoves)
    }

    fun move(fps: Long) {
        if (xVelocity != 0f) {
            worldLocation!!.x += xVelocity / fps
        }
        if (yVelocity != 0f) {
            worldLocation!!.y += yVelocity / fps
        }
    }

    fun checkCollisions(rectHitBoxRight: RectHitBox, rectHitBoxLeft: RectHitBox,
                        rectHitBoxTop: RectHitBox, rectHitBoxBottom: RectHitBox): Int {
        var collided = 0
        if (this.rectHitBoxLeft.intersects(rectHitBoxRight)) {
            setWorldLocationX(rectHitBoxRight.right)
            collided = 1
        }
        if (this.rectHitBoxRight.intersects(rectHitBoxLeft)) {
            setWorldLocationX(rectHitBoxLeft.left - width)
            collided = 1
        }
        if (this.rectHitBoxBottom.intersects(rectHitBoxTop)) {
            setWorldLocationY(rectHitBoxTop.top - height)
            collided = 2
        }
        if (this.rectHitBoxTop.intersects(rectHitBoxBottom)) {
            setWorldLocationY(rectHitBoxBottom.bottom)
            collided = 2
        }
        return collided
    }

    fun setWorldLocationY(y: Float) {
        worldLocation!!.y = y
    }

    fun setWorldLocationX(x: Float) {
        worldLocation!!.x = x
    }

    fun setAnimFps(animFps: Int) {
        this.animFps = animFps
    }

    fun setAnimFrameCount(animFrameCount: Int) {
        this.animFrameCount = animFrameCount
    }

    fun getxVelocity(): Float {
        return xVelocity
    }

    fun getyVelocity(): Float {
        return yVelocity
    }

    fun setxVelocity(xVelocity: Float) {
        if (isMoves) this.xVelocity = xVelocity
    }

    fun setyVelocity(yVelocity: Float) {
        if (isMoves) this.yVelocity = yVelocity
    }
}